README.md
[![Master Build Status](https://img.shields.io/travis/IMSGlobal/caliper-python.svg?label=master)](https://travis-ci.org/IMSGlobal/caliper-python)
[![Develop Build Status](https://img.shields.io/travis/IMSGlobal/caliper-python/develop.svg?label=develop)](https://travis-ci.org/IMSGlobal/caliper-python/develop)


# IMS Global Learning Consortium, Inc.

The Caliper Analytics® Specification provides a structured approach to describing, collecting and
exchanging learning activity data at scale. Caliper also defines an application programming
interface (the Sensor API™) for marshalling and transmitting event data from instrumented
applications to target endpoints for storage, analysis and use.

caliper-python is a reference implementation of the Sensor API™ written in Python.


## Dependencies

To effectively use this `imsglobal_caliper` bundle, you will need to have Python, pip, and
setuptools installed. This package also depends on several third-party packages:

* aniso8601 -- (3-clause BSD licensed)

* future -- (MIT licensed)

* requests -- (Apache 2 licensed)

* rfc3986 -- (Apache 2 licensed)

To work with the `imsglobal_caliper` bundle (maintain, use, run tests) you'll need to have these
third-party packages in your local Python environment.

### Testing

In test, the `ims_global` package depends on these third-party packages not actually used within
the package code itself:

* pytest and pytest-cov

* responses

* tox


### Which platform

We developed `imsglobal_caliper` primarily using the Python 3.7+ platform; however, we've taken
steps to make it portably usbale with Python 2.7 as well (using the `future` package). Changes to
the code base should ensure the tests run clean under Python 3.7, 2.7, PyPy, and PyPy3.


## Build and install

We built `imsglobal_caliper` to be packagable for loading from PyPi, or direct from a source
bundle, with `pip`.

If it's on PyPi, you can simply install it into your site-packages with:

``` shell
pip install imsglobal_caliper
```


### Building and installing from a source bundle

If you want to install it from a source bundle, then use these steps (note that
the name of the repository on GitHub is different to the distribution package name):

1. Clone the [repository from GitHub](https://github.com/IMSGlobal/caliper-python.git>)
   onto your local machine.

2. Use pip to install the package as a writable source bundle; there's no need
   to do a package build step ahead of time:

   ``` shell
   pip3 install -e caliper-python
   ```


## Changing and testing

All the `imsglobal_caliper` code you would use to build and use a caliper sensor in your
application you can find in the `caliper` main module.

**Branches and tags**. This project organizes around two, long-running branches:

- *master*. Stable, deployable branch that stores the official release history.

- *develop*. Unstable development branch; current work that targets a future release is merged to
  this branch.

Project releases are tagged and versioned using git tags of the form
*MAJOR*.*MINOR*.*PATCH*[-*label*] (for example: `1.1.0`; or, for signalling a pre-release version
with a label extension, `1.2.0-RC01`).

**Making contributions**. We welcome the posting of issues by those who are not members of IMS
Global Learning Consortium (for example, for feature requests, bug reports, questions, and so on)
but we do not accept contributions in the form of pull requests from non-members. See
`CONTRIBUTING.md` for more information.

**Testing**. When you work with the source bundle, we include tests in the `tests` directory. In
general, whenever you might make changes to code in `caliper`, you should run the fixture
tests. From the source repo's top-level directory, use `tox` to run all the tests on Python 2.7,
Python 3.7, PyPy, and PyPy3.

Before you can run the tests, you will also need a copy of the Caliper fixtures (see below).

**Fixtures**. The test suites are principally designed to test against the canonical common JSON
fixtures. To set up your tests, you should clone the
[Caliper common fixtures repository](https://github.com/IMSGlobal/caliper-common-fixtures)
repository into the `tests/fixtures` directory, and pull changes there.

Ideally, if you make changes to `imsglobal_caliper` you should ensure that your changes will test
well against the fixtures from the main common fixtures repository. When IMS makes updates to the
`imsglobal_caliper` repository, it ensures that the updates do test well against the common
fixtures repository.


## Using the package

We made `imsglobal_caliper` with a lean integration layer for your application. To use the package,
your application needs only:

``` python
import caliper
```

Your application will need awareness of these parts of the package:

* The `caliper.HttpOptions` class (for use with simple HTTP transport to a Caliper endpoint).

* The `caliper.SimpleSensor` class and its API.

* The `Event` and `Entity` subclasses (found in `caliper.events` and `caliper.entities`,
  respectively) that you will need to use for the metric profile actions you want to support (see
  `caliper.constants` for the association of entities and events with profiles).

Here is a very simple example code scrap that demonstrates how an application might send a basic
navigation event to a caliper endpoint:

``` python
import caliper

the_config = caliper.HttpOptions(
    host='http://caliper-endpoint.your-school.edu/events/',
    auth_scheme='Bearer',
    api_key='your-caliper-API-key' )

# Here you build your sensor; it will have one client in its registry,
# with the key 'default'.
the_sensor = caliper.build_simple_sensor(
    sensor_id = 'http://learning-app.your-school.edu/sensor',
    config_options = the_config )

# Here, you will have caliper entity representations of the various
# learning objects and entities in your wider system, and you provide
# them into the constructor for the event that has just happened.
#
# Note that you don't have to pass an action into the constructor because
# the NavigationEvent only supports one action, part of the
# Caliper base profile: caliper.constants.BASE_PROFILE_ACTIONS['NAVIGATED_TO']
#
the_event = caliper.events.NavigationEvent(
    actor = the_user_currently_acting_as_caliper_Actor_entity,
    edApp = your_application_as_caliper_SoftwareApplication_entity,
    group = the_course_offering_in_play_as_caliper_Organization_entity,
    event_object = the_caliper_DigitalResource_the_actor_is_using,
    referrer = the_caliper_DigitalResource_the_actor_came_from,
    target = the_caliper_DigitalResource_the_actor_is_going_to,
    eventTime = the_time_when_the_actor_did_the_action )

# Once built, you can use your sensor to send one or more often used
# entities; suppose for example, you'll be sending a number of events
# that all have the same actor

sent_identities = the_sensor.send(the_event.actor)

# The return structure from the sensor will be a dictionary of lists: each
# item in the dictionary has a key corresponding to a client key,
# so ret['default'] fetches back the list of URIs of all the @ids of
# the fully described Caliper objects you have sent with that describe call.
#
# Now you can use this list with event sendings to send only the identifiers
# of already-described entities, and not their full forms:

the_sensor.send(the_event, described_objects=sent_identities)

# You can also just send the event in its full form, with all fleshed out
# entities:

the_sensor.send(the_event)

# You can check the status code sent back by the endpoint for the last
# invocation of send():

assert the_sensor.status_code in [200, 201, 202]

# If you create your configuration with debug=true, then your sensor will
# keep a list of the full responses (it uses the "requests" library under the
# covers, so these will be response objects from that library:

the_sensor.config.DEBUG = true
sent_identifies = the_sensor.send(the_event.actor)
the_sensor.send(the_event, described_objects=sent_identities)
responses = the_sensor.debug
```

Your actual use of the caliper code will certainly be more complex than this. For assistance
getting from this very simple example through to more complex and realistic code-use, we encourage
you to look at the unit tests in the package, and the common fixtures they test against.


## Copyright and License

This project is licensed under the terms of the GNU Lesser General Public License (LGPL),
version 3. See the LICENSE file for details. For additional information on licensing options for
IMS members, please see the NOTICE file.

©2019 IMS Global Learning Consortium, Inc. All Rights Reserved.  Trademark Information -
http://www.imsglobal.org/copyright.html


HISTORY.md
# Change history

## 1.2.0

- Supports IMS Caliper specification 1.2.

- Add support for user-agent type information within SoftwareApplication entities, and add new
  optional `client` field in Session entities.

- Simplify implementation by removing marker types, LearningContext entity, as not in the Caliper
  specification.


## 1.1.10

- Add in support for the Feedback metric profile.

  *1.1.10.1*. Bug fix for [Issue #162](https://github.com/IMSGlobal/caliper-python/issues/162).

  *1.1.10.2*. Bug fix for [Issue #168](https://github.com/IMSGlobal/caliper-python/issues/168).

  *1.1.10.3*. Bug fix for [Issue #170](https://github.com/IMSGlobal/caliper-python/issues/170).

  *1.1.10.4*. Bug fix for [Issue #167](https://github.com/IMSGlobal/caliper-python/issues/167).

  *1.1.10.5*. Bug fix for [Issue #176](https://github.com/IMSGlobal/caliper-python/issues/176).

  *1.1.10.6*. Bug fix for [Issue #187](https://github.com/IMSGlobal/caliper-python/issues/187),
              Bug fix for [Issue #183](https://github.com/IMSGlobal/caliper-python/issues/185),
              Bug fix for [Issue #183](https://github.com/IMSGlobal/caliper-python/issues/183),
              Bug fix for [Issue #181](https://github.com/IMSGlobal/caliper-python/issues/181).


## 1.1.9

- Ensure that all Events and Entities have appropriate properties available for
  their expected fields.

  *1.1.9.1*. Bug fix for [Issue #147](https://github.com/IMSGlobal/caliper-python/issues/147).


## 1.1.8

- Provide a SimpleSensor that reduces complexity for most straightforward use
  of the reference implementation: the simple sensor has only the `send()`
  method, used to send both Caliper Entities and Events, and supports only a
  single `HttpRequestor` transport connection to a single configured endpoint.


## 1.1.7

- Add in support for the Research Management metric profile.


## 1.1.6

- No need any longer to import with_metaclass from future.utils, so pruned.


## 1.1.5

- Future package at 0.17.0 has problems when pip attempts to install for both
  Py2 and Py3 local environments; configure tox to enforce no-cache-dir for pip
  when building virtual envs for tests.


## 1.1.4

- Cope with deprecations in dependencies: rfc3986.api.is_valid_uri(s) now
  deprecated in favour of using rfc3986.validators.Validator class; collections
  ABCs no longer loadable from root module in 3.7+, should be loaded from
  collections.abc instead.


## 1.1.3

- Add in a debug option for sensor clients to capture HTTP response objects for
  later inspection.


## 1.1.2

- Fix bug in HttpRequestor in which it could send a malformed HTTP header when
  the requestor's bound HttpOptions had no auth scheme specified


## 1.1.1

- Revise test utility code for new arrangement of fixtures repository
- Revise package constants (including actions, events) for better profile-based
  context handling, including support for the new profile-specific context
  documents
- Export CALIPER_VERSION as top-level package property
- Refactor context examination so that the Caliper Basic profile's context is
  always a valid base for the other, known Caliper profile contexts
- Refactor to move profile/event/action verification into the BaseEvent class,
  and ensure it properly produces the right exception description string on
  unknown actions passed in
- Re-order tox's building of virtual environments to avoid dependency problems
  with pytest/funcsigs
- Improve on cleanliness of test erroring on fixtures not rebuildable as
  Caliper things
- Improve condensor to support more strict context checking
- Add CONTRIBUTING and HISTORY files to package


## 1.1.0

- Supports IMS Caliper specification 1.1.

  *1.1.0.1*. Bug fix for [Issue #147](https://github.com/IMSGlobal/caliper-python/issues/147>).

  *1.1.0.2*. Update of packaging and test framework for CI.


## 1.0.0

- Supports IMS Caliper specification 1.0


caliper/client.py
import collections
import time
from datetime import datetime, timedelta
import json
import logging
import numbers
import threading

from dateutil.tz import tzutc
import requests

from stats import Statistics
from errors import ApiError
from utils import guess_timezone, DatetimeSerializer

import options


logging_enabled = True
logger = logging.getLogger('caliper')


def log(level, *args, **kwargs):
    if logging_enabled:
        method = getattr(logger, level)
        method(*args, **kwargs)


def process_exception(client, data, e):
    log('warn', 'Caliper request error', exc_info=True)
    client._on_failed_flush(data, e)


def process_response(client, data, response):

    if response.status_code == 200 or response.status_code == 201:
        client._on_successful_flush(data, response)
    elif response.status_code == 400:
        content = response.text
        try:
            body = json.loads(content)

            code = 'bad_request'
            message = 'Bad request'

            if 'error' in body:
                error = body.error

                if 'code' in error:
                    code = error['code']

                if 'message' in error:
                    message = error['message']

            client._on_failed_flush(data, ApiError(code, message))

        except Exception:
            client._on_failed_flush(data, ApiError('Bad Request', content))
    else:
        client._on_failed_flush(data,
                                ApiError(response.status_code, response.text))


def request(client, url, data):

    log('debug', 'Request ... with data ' + json.dumps(data))
    
    # Separate out describes
    describes = [x for x in data['batch'] if x['__action'] == 'describe']
    log('debug', 'Describes in list =  ' + json.dumps(describes))
    describe_url = options.host + options.endpoints['describe']
    
    # Separate out measures
    measures = [x for x in data['batch'] if x['__action'] == 'measure']
    log('debug', 'Measures in list =  ' + json.dumps(measures))
    measure_url = options.host + options.endpoints['measure']

    # Every thing else
    defaults = [x for x in data['batch'] if x['__action'] == 'identify' or x['__action'] == 'alias' or x['__action'] == 'track']
    log('debug', 'Defaults in list =  ' + json.dumps(defaults))

    try:

        response_flag = True

        if len(describes) > 0:
            for describe in describes:
                response_d = requests.put(describe_url,
                                         data=json.dumps(describe, cls=DatetimeSerializer),
                                         headers={'content-type': 'application/json'},
                                         timeout=client.timeout)
                log('debug', 'Finished Caliper describe request. Response = ' + response_d.text)
                process_response(client, describe, response_d)
                response_flag = response_flag and (response_d.status_code == 200 or response_d.status_code == 201)

        if len(measures) > 0:
            for measure in measures:
                response_m = requests.put(measure_url,
                                         data=json.dumps(measure, cls=DatetimeSerializer),
                                         headers={'content-type': 'application/json'},
                                         timeout=client.timeout)
                log('debug', 'Finished Caliper measure request. Response = ' + response_m.text)
                process_response(client, measure, response_m)
                response_flag = response_flag and (response_m.status_code == 200 or response_m.status_code == 201)

        if len(defaults) > 0:
            response = requests.put(url,
                                 data=json.dumps(defaults, cls=DatetimeSerializer),
                                 headers={'content-type': 'application/json'},
                                 timeout=client.timeout)

            log('debug', 'Finished Caliper describe request. Response = ' + response.text)
            process_response(client, data, response)
            response_flag = response_flag and (response.status_code == 200 or response.status_code == 201)

        log('debug', 'Request - response_flag is ' + str(response_flag))

        return response_flag

    except requests.ConnectionError as e:
        process_exception(client, data, e)
    except requests.Timeout as e:
        process_exception(client, data, e)

    return False


class FlushThread(threading.Thread):

    def __init__(self, client):
        threading.Thread.__init__(self)
        self.client = client

    def run(self):
        log('debug', 'Flushing thread running ...')

        self.client._sync_flush()

        log('debug', 'Flushing thread done.')


class Client(object):
    """The Client class is a batching asynchronous python wrapper over the
    Caliper API.

    """

    def __init__(self, secret=None, log_level=logging.INFO, log=True,
                 flush_at=20, flush_after=timedelta(0, 10),
                 async=True, max_queue_size=10000, stats=Statistics(),
                 timeout=10, send=True):
    
        """Create a new instance of a caliper-python Client

        :param str secret: The Caliper API secret
        :param logging.LOG_LEVEL log_level: The logging log level for the
        client talks to. Use log_level=logging.DEBUG to troubleshoot
        : param bool log: False to turn off logging completely, True by default
        : param int flush_at: Specicies after how many messages the client will
        flush to the server. Use flush_at=1 to disable batching
        : param datetime.timedelta flush_after: Specifies after how much time
        of no flushing that the server will flush. Used in conjunction with
        the flush_at size policy
        : param bool async: True to have the client flush to the server on
        another thread, therefore not blocking code (this is the default).
        False to enable blocking and making the request on the calling thread.
        : param float timeout: Number of seconds before timing out request to
        Caliper
        : param bool send: True to send requests, False to not send. False to
        turn caliper off (for testing).
        """

        self.secret = secret

        self.queue = collections.deque()
        self.last_flushed = None

        if not log:
            # TODO: logging_enabled is assigned, but not used
            logging_enabled = False
            # effectively disables the logger
            logger.setLevel(logging.CRITICAL)
        else:
            logger.setLevel(log_level)

        self.async = async

        self.max_queue_size = max_queue_size
        self.max_flush_size = 50

        self.flush_at = flush_at
        self.flush_after = flush_after

        self.timeout = timeout

        self.stats = stats

        self.flush_lock = threading.Lock()
        self.flushing_thread = None

        self.send = send

        self.success_callbacks = []
        self.failure_callbacks = []

        self.current_time = lambda: int(round(time.time() * 1000))

    def set_log_level(self, level):
        """Sets the log level for caliper-python

        :param logging.LOG_LEVEL level: The level at which caliper-python log
        should talk at
        """
        logger.setLevel(level)

    def _check_for_secret(self):
        if not self.secret:
            raise Exception('Please set caliper.secret before calling ' +
                            'identify or track.')

    def _coerce_unicode(self, cmplx):
        return unicode(cmplx)

    def _clean_list(self, l):
        return [self._clean(item) for item in l]

    def _clean_dict(self, d):
        data = {}
        for k, v in d.iteritems():
            try:
                data[k] = self._clean(v)
            except TypeError:
                log('warn', 'Dictionary values must be serializeable to ' +
                            'JSON "%s" value %s of type %s is unsupported.'
                            % (k, v, type(v)))
        return data

    def _clean(self, item):
        if isinstance(item, (str, unicode, int, long, float, bool,
                             numbers.Number, datetime)):
            return item
        elif isinstance(item, (set, list, tuple)):
            return self._clean_list(item)
        elif isinstance(item, dict):
            return self._clean_dict(item)
        else:
            return self._coerce_unicode(item)

    def on_success(self, callback):
        """
        Assign a callback to fire after a successful flush

        :param func callback: A callback that will be fired on a flush success
        """
        self.success_callbacks.append(callback)

    def on_failure(self, callback):
        """
        Assign a callback to fire after a failed flush

        :param func callback: A callback that will be fired on a failed flush
        """
        self.failure_callbacks.append(callback)

    def describe(self, entity_type=None, entity_id=None, properties={}, timestamp=None):

        log('debug', 'Client - describe')

        self._check_for_secret()

        if not entity_type:
            raise Exception('Must supply a entity_type.')

        if not entity_id:
            raise Exception('Must supply a entity_id.')

        if properties is not None and not isinstance(properties, dict):
            raise Exception('properties must be a dictionary.')

        if timestamp is None:
            timestamp = self.current_time()
        # elif not isinstance(timestamp, datetime):
        #     raise Exception('Timestamp must be a datetime object.')
        # else:
        #     timestamp = guess_timezone(timestamp)

        action = {'entityId':    entity_id,
                  'type':        entity_type,
                  'properties':  properties,
                  'timestamp':   timestamp,
                  'apiKey':      self.secret,
                  '__action':      'describe'}

        if self._enqueue(action):
            self.stats.describes += 1

    def measure(self, action=None, learning_context={}, activity_context={}, timestamp=None):

        log('debug', 'Client - measure')

        self._check_for_secret()

        if not action:
            raise Exception('Must supply an action.')

        if learning_context is not None and not isinstance(learning_context, dict):
            raise Exception('learning_context must be a dictionary.')

        if activity_context is not None and not isinstance(activity_context, dict):
            raise Exception('activity_context must be a dictionary.')

        if timestamp is None:
            timestamp = self.current_time()
        # elif not isinstance(timestamp, datetime):
        #     raise Exception('Timestamp must be a datetime object.')
        # else:
        #     timestamp = guess_timezone(timestamp)

        action = {'action':           action,
                  'learningContext':  learning_context,
                  'activityContext':  activity_context,
                  'timestamp':        timestamp,
                  'apiKey':           self.secret,
                  '__action':         'measure'}

        if self._enqueue(action):
            self.stats.measures += 1

    def _should_flush(self):
        """ Determine whether we should sync """

        full = len(self.queue) >= self.flush_at
        stale = self.last_flushed is None

        if not stale:
            stale = datetime.now() - self.last_flushed > self.flush_after

        return full or stale

    def _enqueue(self, action):

        # if we've disabled sending, just return False
        if not self.send:
            return False

        submitted = False

        if len(self.queue) < self.max_queue_size:
            self.queue.append(action)

            self.stats.submitted += 1

            submitted = True

            log('debug', 'Enqueued ' + action['__action'] + '.')

        else:
            log('warn', 'caliper-python queue is full')

        if self._should_flush():
            self.flush()

        return submitted

    def _on_successful_flush(self, data, response):
        if 'batch' in data:
            for item in data['batch']:
                self.stats.successful += 1
                for callback in self.success_callbacks:
                    callback(data, response)
        else:
            log('debug', '*** dealing with data ' + json.dumps(data))
            if data['__action'] == 'describe' or data['__action'] == 'measure':
                self.stats.successful += 1
                for callback in self.success_callbacks:
                    callback(data, response)

    def _on_failed_flush(self, data, error):
        if 'batch' in data:
            for item in data['batch']:
                self.stats.failed += 1
                for callback in self.failure_callbacks:
                    callback(data, error)

    def _flush_thread_is_free(self):
        return self.flushing_thread is None \
            or not self.flushing_thread.is_alive()

    def flush(self, async=None):
        """ Forces a flush from the internal queue to the server

        :param bool async: True to block until all messages have been flushed
        """

        flushing = False

        # if the async arg is provided, it overrides the client's settings
        if async is None:
            async = self.async

        if async:
            # We should asynchronously flush on another thread
            with self.flush_lock:

                if self._flush_thread_is_free():

                    log('debug', 'Initiating asynchronous flush ..')

                    self.flushing_thread = FlushThread(self)
                    self.flushing_thread.start()

                    flushing = True

                else:
                    log('debug', 'The flushing thread is still active.')
        else:

            # Flushes on this thread
            log('debug', 'Initiating synchronous flush ..')
            self._sync_flush()
            flushing = True

        if flushing:
            self.last_flushed = datetime.now()
            self.stats.flushes += 1

        return flushing

    def _sync_flush(self):

        log('debug', 'Starting flush ..')

        successful = 0
        failed = 0

        url = options.host + options.endpoints['batch']

        while len(self.queue) > 0:

            batch = []
            for i in range(self.max_flush_size):
                if len(self.queue) == 0:
                    break

                batch.append(self.queue.pop())

            payload = {'batch': batch, 'secret': self.secret}

            if request(self, url, payload):
                successful += len(batch)
            else:
                failed += len(batch)

        log('debug', 'Successfully flushed {0} items [{1} failed].'.
                     format(str(successful), str(failed)))